File: 3rd Semester (C#)/Lab4/Banks/Models/TransferTransaction.cs
using Banks.Interfaces;
using Banks.Tools;

namespace Banks.Models;

public class TransferTransaction : ITransaction
{
    public TransferTransaction(IAccount account_from, IAccount account_to, double moneyAmount, double comission, DateTime dateTime)
    {
        Id = Guid.NewGuid();
        Account = account_from;
        TransferAccount = account_to;
        MoneyAmount = moneyAmount;
        Comission = comission;
        DateTime = dateTime;
    }

    public Guid Id { get; }
    public IAccount Account { get; }
    public IAccount TransferAccount { get; }
    public double MoneyAmount { get; }
    public double Comission { get; }
    public DateTime DateTime { get; }
    public bool Canceled { get; private set; }

    public void Cancel()
    {
        if (Canceled)
        {
            throw new BanksException($"Failed to Cancel transaction: {this}, transaction is already canceled!");
        }

        Account.Money += MoneyAmount + Comission;
        TransferAccount.Money -= MoneyAmount;
        Canceled = true;
    }
}

File: 4th Semester (Java)/lab-1/src/main/java/transactions/TransferTransaction.java
package transactions;

import accounts.Account;
import exceptions.BanksException;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Represents a transaction of transferring money from one account to another.
 */
public class TransferTransaction extends Transaction {
    /**
     * The account money were transferred to.
     */
    public Account transferAccount;

    /**
     * TransferTransaction constructor.
     * @param accountFrom The account money were taken from
     * @param accountTo The account money were taken to
     * @param moneyAmount The amount of money was transferred
     * @param commission The amount of money was taken from user
     * @param dateTime The date and time of transaction
     */
    public TransferTransaction(Account accountFrom, Account accountTo, double moneyAmount, double commission, LocalDateTime dateTime) {
        Id = UUID.randomUUID();
        account = accountFrom;
        this.transferAccount = accountTo;
        this.moneyAmount = moneyAmount;
        this.commission = commission;
        this.dateTime = dateTime;
    }

    /**
     * Cancels transaction, takes money from TransferAccount and returns it to Account.
     * @throws BanksException if transaction is already canceled
     */
    public void cancel() throws BanksException {
        if (canceled) {
            throw new BanksException("Failed to Cancel transaction, transaction is already canceled!");
        }

        double newAccountMoneyAmount = account.getMoney() + moneyAmount + commission;
        double newTransferAccountMoneyAmount = transferAccount.getMoney() - moneyAmount;
        account.setMoney(newAccountMoneyAmount);
        transferAccount.setMoney(newTransferAccountMoneyAmount);
        canceled = true;
    }

    public Account getTransferAccount() {
        return transferAccount;
    }
}
